### ISSUE TITLE
Common: find a way to reduce assembly time

### ISSUE THREAD

**Issue by fblundun** [snowplow/snowplow#1665]

**Comment by fblundun** [comment]
See http://stackoverflow.com/questions/19543530/sbt-assembly-task-runs-slowly-after-adding-some-dependencies

**Comment by fblundun** [comment]
So far I haven't got this to work - I am following [these instructions](https://github.com/sbt/sbt-assembly/blob/5657dacd0325fc7606ace12adadbdcfea22d2d0c/notes/0.10.0.markdown#breaking-changes) and have added this:

```
    assemblyOption in assembly ~= { _.copy(cacheOutput = false) },
```

but without noticeable speedup.

I notice that `sbt assembly` runs much faster in the host than in the guest. Maybe increasing the number of cores we give the snowplow/snowplow VM would help?

**Comment by alexanderdean** [comment]
Hey Fred - giving the VM more cores sounds like a good idea!

**Comment by fblundun** [comment]
Here are some interesting links on giving a VM multiple cores:

http://superuser.com/questions/297697/are-there-any-real-advantages-to-assigning-more-cores-to-virtualbox

https://stefanwrobel.com/how-to-make-vagrant-performance-not-suck

**Comment by alexanderdean** [comment]
Thanks Fred, could you rename this ticket to something like "Update Vagrant to give access to all CPU cores" and close the sibling tickets?

**Comment by fblundun** [comment]
Unfortunately increasing to 4 cores doesn't seem to reduce the assembly time either.

In this thread, joeroot is experiencing the same problem of performance on guest being much worse than performance on host: https://github.com/sbt/sbt-assembly/issues/68

**Comment by fblundun** [comment]
Descheduling since we haven't made any progress.

**Comment by fblundun** [comment]
@alexanderdean @jbeemster I think I've finally made progress on this one. My idea is to do the assembly outside the shared directory - i.e. copy the project to be assembled out of /vagrant and into e.g. /home/vagrant before assembling.

I am testing this with the current version of scala-kinesis-enrich on a VM with 1 core without NFS enabled. Assembling outside the shared directory took 63 seconds. Assembling inside the shared directory is currently at 20 minutes and still going.

**Comment by alexanderdean** [comment]
Genius!

**Comment by jbeemster** [comment]
It really is!

**Comment by jramos** [comment]
@fblundun - eureka! I've got a better idea: change your shared folder type from `vboxfs` (the default) to `nfs` in your Vagrantfile. E.g.

```
config.vm.synced_folder "~/src/", "/home/vagrant/src", type: "nfs", create: true
```

You'll want to setup NFS in your VM, too. It's different depending on your distro and config, but here's what I did:

```bash
sudo su
apt-get install nfs-kernel-server nfs-common
echo "/home/vagrant   127.0.0.1(rw,no_subtree_check,all_squash,async)" | tee -a /etc/exports
service nfs-kernel-server restart
```

The `echo` piece is needed because NFS won't start on Ubuntu unless there's a defined export. So I share `/home/vagrant` with myself.

Assembly time went from ~8 minutes to ~10 seconds for me. :-)
